library/src/main/java/com/q42/q42stats/library/Q42Stats.kt

```diff
@@ -1,5 +1,74 @@
 package com.q42.q42stats.library
 
-class Q42Stats {
-    val name = "Q42 Stats lib"
+import android.content.Context
+import com.q42.q42stats.library.collector.AccessibilityCollector
+import com.q42.q42stats.library.collector.PreferencesCollector
+import com.q42.q42stats.library.collector.SystemCollector
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import java.io.Serializable
+import java.util.concurrent.atomic.AtomicBoolean
+
+const val TAG = "Q42Stats"
+
+private val MUTEX = Unit
+
+class Q42Stats(private val config: Q42StatsConfig) {
+    private val isRunning = AtomicBoolean(false)
+
+    /* Collects stats and sends it to the server. This method is safe to be called from anywhere
+    in your app and will do nothing if it has already run before */
+    fun runAsync(context: Context) {
+        Q42StatsLogger.d(TAG, "Q42Stats: Checking Preconditions")
+        if (isRunning.get()) {
+            Q42StatsLogger.i(
+                TAG,
+                "Q42Stats is already running. Exit."
+            )
+            return
+        }
+        GlobalScope.launch(Dispatchers.IO) { synchronized(MUTEX) { runSync(context) } }
```

Comment:
multiple threads might reach here, and then start waiting, because you don't getAndSet

Reply:
The easiest to read for me (and less error prone) would be: leave the Atomics and Synchronized blocks out and go on "Dispatchers.Main" (always single threaded, that makes code very readable and not error prone). Then: check if "val job" (inside your class) is not null or !isActive(), if so: start a new one now (else do nothing). Switch to another dispatcher inside that job to do your heavy work.

Reply:
I think you made a mistake in your suggested logic, but did implement the idea. Let's discuss further in my new PR